src/libraries/NormalizedAmount.sol

```diff
@@ -1,58 +1,171 @@
+    }
+
+    function add(
+        NormalizedAmount memory a,
+        NormalizedAmount memory b
+    ) internal pure returns (NormalizedAmount memory) {
+        // on initialization
+        if (isZero(a)) {
+            return b;
+        }
+
+        if (isZero(b)) {
+            return a;
         }
-        // amount after normalization must fit into uint64 (that's the point of
-        // normalization, as Solana only supports uint64 for token amounts)
-        if (amount > type(uint64).max) {
-            revert AmountTooLarge(amount);
+
+        if (a.decimals != b.decimals) {
+            revert NumberOfDecimalsNotEqual(a.decimals, b.decimals);
         }
-        return NormalizedAmount.wrap(uint64(amount));
+        return NormalizedAmount(a.amount + b.amount, a.decimals);
     }
 
-    function denormalize(NormalizedAmount amount, uint8 decimals) internal pure returns (uint256) {
-        uint256 denormalized = NormalizedAmount.unwrap(amount);
-        if (decimals > 8) {
-            denormalized *= 10 ** (decimals - 8);
+    function min(
+        NormalizedAmount memory a,
+        NormalizedAmount memory b
+    ) public pure returns (NormalizedAmount memory) {
+        // on initialization
+        if (isZero(a) && !isZero(b)) {
+            return a;
+        }
+        if (isZero(b) && !isZero(a)) {
+            return b;
+        }
+
+        if (a.decimals != b.decimals) {
+            revert NumberOfDecimalsNotEqual(a.decimals, b.decimals);
         }
-        return denormalized;
+
+        return a.amount < b.amount ? a : b;
+    }
+
+    function minDecimals(uint8 toDecimals, uint8 fromDecimals) internal pure returns (uint8) {
+        return toDecimals < fromDecimals ? toDecimals : fromDecimals;
+    }
+
+    /// @dev scale the amount from origDecimals to normDecimals (base 10)
+    function scalingFactor(
+        uint8 origDecimals,
+        uint8 normDecimals
+    ) internal pure returns (uint256) {
+        if (origDecimals > normDecimals) {
+            return 10 ** (origDecimals - normDecimals);
+        } else {
+            return 1;
+        }
+    }
+
+    function normalize(
+        uint256 amt,
+        uint8 fromDecimals
+    ) internal pure returns (NormalizedAmount memory) {
+        uint8 toDecimals = minDecimals(NORMALIZED_DECIMALS, fromDecimals);
+        amt /= scalingFactor(fromDecimals, toDecimals);
+
+        // NOTE: amt after normalization must fit into uint64 (that's the point of
+        // normalization, as Solana only supports uint64 for token amts)
+        if (amt > type(uint64).max) {
+            revert AmountTooLarge(amt);
+        }
+        return NormalizedAmount(uint64(amt), toDecimals);
+    }
+
+    function denormalize(
+        NormalizedAmount memory amt,
+        uint8 toDecimals
+    ) internal pure returns (uint256) {
+        (uint256 deNorm, uint8 dec) = unwrap(amt);
+        deNorm *= scalingFactor(toDecimals, dec);
```

Comment:
Correct me if I'm wrong but I don't think this denormalises correctly when `toDecimals` < 8 and `fromDecimals` >= 8?

For example if normalise 1e8 of a token with 8 decimals on the source chain we'd have (1e8, 8) in the wire format. Then when we try to denormalise on the destination chain to `toDecimals = 6` the scaling factor is set to 1 and the amount out is 1e8. So we end up getting 100 tokens on the destination chain instead of 1

Reply:
another way to say this is that denormalization strictly depends on normalization.

Reply:
What if we have the following case:
- Token on Eth -- 18 decimals
- Token on Arbitrum -- 6 decimals
Token on Eth will get normalized to 8 decimals and transferred over. Then on the Arbitrum side, the 8 decimals will get scaled by 1 and remain 8 decimals so we get 100x more tokens as @djb15 is saying.

Reply:
Yeah my original message is overly constrained as well. I think this is broken when `toDecimals` < `normalised(fromDecimals)`